test/pulses/cos_square_pulse.jl

```diff
@@ -0,0 +1,62 @@
+using Random
+using IntervalSets
+using QEDfields
+using QEDbase
+using QuadGK
+
+RNG = MersenneTwister(123456789)
+ATOL = 0.0
+RTOL = sqrt(eps())
+
+DPHIS = [rand(RNG), rand(RNG) * 10, rand(RNG) * 100, rand(RNG) * 1000, rand(RNG), 10000]
+
+# wrapper implementation to test analytical solutions of the generic spectrum
+
+struct CosSquarePulseWrapper{C<:CosSquarePulse} <: AbstractPulsedPlaneWaveField
+    pulse::C
+end
+QEDfields.reference_momentum(p::CosSquarePulseWrapper) = reference_momentum(p.pulse)
+QEDfields.domain(p::CosSquarePulseWrapper) = domain(p.pulse)
+QEDfields.phase_duration(p::CosSquarePulseWrapper) = phase_duration(p.pulse)
+QEDfields._envelope(p::CosSquarePulseWrapper, x) = QEDfields._envelope(p.pulse, x)
+
+@testset "pulse interface" begin
+    @test hasmethod(reference_momentum, Tuple{CosSquarePulse})
+    @test hasmethod(domain, Tuple{CosSquarePulse})
+    @test hasmethod(phase_duration, Tuple{CosSquarePulse})
+    @test hasmethod(QEDfields._envelope, Tuple{CosSquarePulse,Real})
+end
+@testset "dphi: $dphi" for dphi in DPHIS
+    test_mom = rand(RNG, SFourMomentum)
+    test_pulse = CosSquarePulse(test_mom, dphi)
+
+    @testset "properties" begin
+        @test reference_momentum(test_pulse) == test_mom
+        @test domain(test_pulse) == Interval(-dphi, dphi) # cos_square specific
+        @test phase_duration(test_pulse) == dphi
+    end
+
+    @testset "envelope" begin
+        # unity at the origin
+        @test envelope(test_pulse, 0.0) == 1.0
+
+        # zero at the endpoints
+        @test isapprox(envelope(test_pulse, -Inf), 0.0, atol=ATOL, rtol=RTOL)
+        @test isapprox(envelope(test_pulse, Inf), 0.0, atol=ATOL, rtol=RTOL)
+    end
+    @testset "generic spectrum" begin
+        wrapper_pulse = CosSquarePulseWrapper(test_pulse)
+        test_pnums = [1.0, -1.0, 1 + rand(RNG) * 0.1, -1 - rand(RNG) * 0.1]
+        @testset "pnum: $pnum" for pnum in test_pnums
+            test_val_xpol = generic_spectrum(test_pulse, PolX(), pnum)
+            test_val_ypol = generic_spectrum(test_pulse, PolY(), pnum)
+
+            groundtruth_xpol = generic_spectrum(wrapper_pulse, PolX(), pnum)
+            groundtruth_ypol = generic_spectrum(wrapper_pulse, PolY(), pnum)
```

Comment:
I don't understand in which way this groundtruth can possibly differ from the test_pulse, except for the case when you wrongly define the method forwarding in lines 18-21 in this file. but this is a problem of the test, not of the `CosSquarePulse` implementation.

Reply:
Implementing the wrapper type (lines 15-21) dispatches on the generic fallback instead of the specialized version. For instance, 

```Julia 
generic_spectrum(CosSquarePulse(K,delta_phi), PolX(), pnum)
```
calls the special implementation `src/pulses/cos_square_pulse.jl:27`, but 
```Julia 
generic_spectrum(CosSquarePulseWrapper(K,delta_phi), PolX(), pnum)
```
calls the generic implementation `src/interfaces/background_field_interface.jl:188`. Therefore, the idea of this construction is to test the special implementation against the generic one. This should also work for other pulse shapes, which means the file [test/pulses/cos_square_pulse.jl](https://github.com/QEDjl-project/QEDfields.jl/pull/24/files/19385dd7aa5c672b690ec5a99aa6cf8a7dccdc76#diff-7623933b2a6ce2dc25ede7792aee43fc34a8939cd94b11a2e3d6687ec7ea87a7) could be generalized easily if other pulses are implemented.